### ISSUE TITLE
Should be possible to specify that "main" include should be at the top

### ISSUE THREAD

**Issue by arneforlie** [uncrustify/uncrustify#2160]
In #2092 we found out that the include_category_[123] feature does not support a feature simlar to clang format's IncludeIsMainRegex.
The feature description is pasted here for reference:
```
IncludeIsMainRegex (std::string)

    Specify a regular expression of suffixes that are allowed in the file-to-main-include mapping.

    When guessing whether a #include is the “main” include (to assign category 0, see above),
    use this regex of allowed suffixes to the header stem. A partial match is done, so that: - “”
    means “arbitrary suffix” - “$” means “no suffix”

    For example, if configured to “(_test)?$”, then a header a.h would be seen as the “main”
    include in both a.cc and a_test.cc.
```
See also #1929

**Comment by mwoehlke-kitware** [comment]
Yeah, this is pretty much a subset of #1929... Note that clang's mechanism is inadequate; we also want to be able to match prefixes (e.g. Qt sources have stuff like "ui_Foo.h", "moc_Foo.cpp").

Upon (brief) additional thought, I think the way this should work is to have two options. The first is a regex that "matches" the file name, and exists to create capture groups. The second is a regex that matches "associated headers"; the second can use capture expressions (i.e. `\1`), which are replaced with the captures from the first regex prior to matching.

So, for example, if the first is `"f(.*)\.([ch].*)"`, and the second is `"x_\1.h"`, then in "foo.cpp", "x_oo.h" would be a 'main' header.

**Comment by avarga** [comment]
+1

This would allow formatting according to the Google C++ Style Guide: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
